### DISCUSSION TITLE
Is it possible to add grid elements using a loop?

### DISCUSSION THREAD

**Discussion by MordechaiHadad** [AvaloniaUI/Avalonia#10144] [Q&A] [answered]
I am wondering if it's possible to add grid elements using loops, let's say I have this code:
```xml
<Grid ShowGridLines="True" RowDefinitions="*,*" Width="800" ColumnDefinitions="*,*">
    <Rectangle Grid.Row="0" Grid.Column="0" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" Fill="Black" />
    <Rectangle Grid.Row="0" Grid.Column="1" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" Fill="White" />
</Grid>
```

into something like this:
```cs
var grid = new Grid(); 
for(int i = 0; i < some_number; i++) {
    if(i % 2 = 0)
     {
         grid[some_index][some_other_index] = new Rectange(Fill = "White");
    } 
    else 
    {    
        grid[some_index][some_other_index] = new Rectange(Fill = "Black");
    }
} 
```

**Comment by maxkatz6** [comment]
You need to set a property value:
```
new Rectangle
{
    [Grid.RowProperty] = 1
}
```
or
rectangle.SetValue(Grid.RowProperty, 1)

**Reply by MordechaiHadad** [reply]
wait so creating a grid in the way I showed is possible? and if so how would the Rectangle know which grid to be part of?

**Reply by workgroupengineering** [reply]
```csharp
            var grid = new Grid();
            var nRows = 4;
            var nCols = 4;
            var totalItems = nRows * nCols;
            var size = new GridLength(50, GridUnitType.Pixel);
            for (int i = 0; i < nRows; i++)
            {
                grid.RowDefinitions.Add(new() { Height = size });
            }
            for (int i = 0; i < nCols; i++)
            {
                grid.ColumnDefinitions.Add(new() { Width = size });
            }

            for (int i = 0; i < totalItems; i++)
            {
                var row = (i / nRows);
                var col = i - (row * nRows);
                var even_row = (row % 2) == 0;
                var fill = (col % 2 == 0) switch
                {
                    true when even_row => Brushes.White,
                    true when !even_row => Brushes.Black,
                    false when even_row => Brushes.Black,
                    false when !even_row => Brushes.White,
                };
                grid.Children.Add(new Rectangle()
                {
                    [Grid.RowProperty] = row,
                    [Grid.ColumnProperty] = col,
                    [Shape.FillProperty] = fill,
                });
            }

```

**Reply by timunie** [reply]
While it's in theory possible to do this, I wonder why you need to do it. Most of the time it's better to go with the MVVM route and use an `ObservalbeColleciton` to add items dynamically. 

You may then present them in an `ItemsControl` or `ListBox`. Alternating row background can also be done. 

Can you share your use case and the final goal you have?

**Reply by MordechaiHadad** [reply]
@timunie I am basically trying to create a chess board which has like 64 squares, which I want to be able to represent as a grid(this seems like the best solution currently)

**Reply by timunie** [reply]
Ah cool. probably `UniformGrid` would also be an option for you. You don't need to calculate Row and Column in that case.

**Reply by MordechaiHadad** [reply]
I will check it thank you! I just started a learning avalonia a day ago lmao.

**Reply by timunie** [reply]
Welcome. Check also out telegram chat for quick questions: https://t.me/Avalonia

**Reply by MordechaiHadad** [reply]
Will do, but why telegram and not discord?

**Reply by timunie** [reply]
Well, the team prefers telegram for quick questions. For things that needs to stay longer, prefer Discussions as this one.

**Reply by MordechaiHadad** [reply]
I have one last question, where exactly do I put the "UI Code" I thought putting it inside `MainWindow` func inside `MainWindow.axaml.cs`, would do something but doesn't seem like how this works.
So where exactly? I am also using `MVVM` cu this is what I thought was recommended.

Edit 1: I tried setting window's content to the Grid but obviously the content inside the xaml file didnt apply, so is it possible to use both?
Edit 2: Huh it isnt

**Reply by MordechaiHadad** [reply]
Okay lmao I figured it out by using this.FindControl();
problem I have encountered is when building the code won't run, I have this code:

MainWindow.axaml:
```xml
    <Panel>
        <UniformGrid Width="500" Rows="8" Columns="8" x:Name="Test"></UniformGrid>
    </Panel>
```
MainWindow.axaml.cs:
```cs
    public MainWindow()
    {
        InitializeComponent();

        var grid = this.FindControl<UniformGrid>("Test");

        grid!.Children.Add(new Rectangle { [Shape.FillProperty] = "White" });
    }
```
    
 shouldn't this work?

**Reply by timunie** [reply]
It will not convert from string automatically. Must be a Brush:

```cs
var grid = this.FindControl<UniformGrid>("Test");
grid!.Children.Add(new Rectangle { [Shape.FillProperty] = Brushes.MyColor , Width = 20, Height = 20});
```

**Reply by MordechaiHadad** [reply]
o, Thank you!

**Comment by jetspiking** [comment]
@MordechaiHadad Avalonia solves this similarly to WPF.

You can create a `ColumnDefinition` and `RowDefinition` programatically (add this to the grid). 

`Grid.RowDefinitions` and `Grid.ColumnDefinitions` are of type `IList` (list), so you can simply add the row and column to it using `list.Add()`.

For specifying the row definitions:
https://reference.avaloniaui.net/api/Avalonia.Controls/RowDefinitions/

For specifying the column definitions:
https://reference.avaloniaui.net/api/Avalonia.Controls/ColumnDefinitions/

Afterwards you can specify the `RowProperty` and `ColumnProperty` for the added child element.

For specifying the row:
https://reference.avaloniaui.net/api/Avalonia.Controls/Grid/9A66BFA3 

For specifying the column:
https://reference.avaloniaui.net/api/Avalonia.Controls/Grid/21D3B58D

**Reply by MordechaiHadad** [reply]
I will check it out thank you!

**Comment by timunie** [comment, marked as answer]
Okay, I found that this could fun to try on my own. So I've build a flexible sample on my own using the MVVM approach. 

Result: 
![image](https://user-images.githubusercontent.com/47110241/216320990-915f0111-51c2-4c49-bfa8-d1fcff533c3b.png)

Source: https://github.com/timunie/Tims.Avalonia.Samples/tree/main/src/ChessBoardSample 


Happy coding

Tim

**Reply by MordechaiHadad** [reply]
Damn that looks actually pretty simple thank you!

**Reply by XDgl233** [reply]
I am working on something similar, a 4x10 grid stash like those from ARPG games, and I want to add items with different grid sizes (some small items are 1x1 in size, while some are 2x3 for example). So is it possible to set Columnspan similarly in Grid? I tried but seems not possible. Or can I achieve the goal via Grid as ItemsControl.ItemsPanel? Thx :)

**Reply by timunie** [reply]
You can set Grid.ColumnSpan in the same way
